### PR TITLE
Script to probe the nccl libraries that PyTorch uses

### LINKED ISSUE #252: PyTorch-based utility to recommend env vars.
Need a PyTorch script to probe instances and software stack, then prints the necessary EFA env vars. Use this script to avoid spending hours to debug runtime crash caused by "forgetting to set env var XXX when using a combination of `libXXX-versionA` and `libYYY-versionB`

It can be a plain Python script, as long as it can accurately probe the version of nccl, aws-ofi-nccl (nccl-net), libfabric, and efa version (dlopen or what not). The probing mechanic must be robust to whether the script is run under container (vs) host, whether env vars are overriden by job script, etc.

At the very least, the script must covers the env vars in EFA cheatsheet, +new ones (like the cuda mem sync env var for incompatibility between specific versions of nccl and efa installer).

### PR DESCRIPTION
**Pull request by verdimrc**
*Issue #, if available:* close #252

*Description of changes:* Probe what PyTorch actually uses for the nccl stacks.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by perifaws** [comment]
cancel or do we move forward with it?

**Comment by verdimrc** [comment]
Ready for review.

File: 4.validation_and_observability/6.probe-torch-nccl-aws-libs/README.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
+# Runtime sanity checks: which NCCL is loaded by PyTorch
+
+The scripts in this folder disambiguate the exact NCCL libraries that a PyTorch application actually
+uses, in the presence of potentially multiple installed versions.
+
+## 1. Motivation
```
Comment (sean-smith) [inline 4.validation_and_observability/6.probe-torch-nccl-aws-libs/README.md:6]: Include usage at the top

File: 4.validation_and_observability/6.probe-torch-nccl-aws-libs/README.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+     47352:       trying file=/fsx/marcverd/awsome-distributed-training/3.test_cases/10.FSDP/conda_env_pytorch/lib/python3.10/site-packages/torch/lib/../../nvidia/nccl/lib/libnccl.so.2
+     47352:     calling init: /fsx/marcverd/awsome-distributed-training/3.test_cases/10.FSDP/conda_env_pytorch/lib/python3.10/site-packages/torch/lib/../../nvidia/nccl/lib/libnccl.so.2
+torch.cuda.nccl.version()=(2, 19, 3)
+     47352:     calling fini: /fsx/marcverd/awsome-distributed-training/3.test_cases/10.FSDP/conda_env_pytorch/lib/python3.10/site-packages/torch/lib/../../nvidia/nccl/lib/libnccl.so.2 [0]
+```
+
+By now, hopefully you're convinced on the need for a runtime probe to pinpoint the version of NCCL
+loaded by PyTorch.
+
+## 2. Howto
```
Comment (sean-smith) [inline 4.validation_and_observability/6.probe-torch-nccl-aws-libs/README.md:83]: How to

File: 4.validation_and_observability/6.probe-torch-nccl-aws-libs/README.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+- one GPU.
+- the command `python` will invoke the binary from your PyTorch environment. Typically, you must
+  activate your environment (conda, virtual env, etc.).
+
+Direct invocation: `./probe-pt-nccl-aws-libs.sh`
+
+Via Slurm: `srun -l -N1 ./probe-pt-nccl-aws-libs.sh`
+
+```console
+$ srun -l -N1 ./probe-pt-nccl-aws-libs.sh
```
Comment (sean-smith) [inline 4.validation_and_observability/6.probe-torch-nccl-aws-libs/README.md:96]: Can this be used as a library within another script? i.e. I want to use it in my `efa-versions.py` to grab the nccl version but invoking the shell script and grep-ing the output is less than ideal.

**Review by sean-smith — COMMENTED** [review]
I'd like to see this give more structured data i.e allow it to be used within other python scripts.

**Comment by verdimrc** [comment]
> I'd like to see this give more structured data i.e allow it to be used within other python scripts

my quick 2c: for this purpose, it's best to re-implement the logic in pure python. Because the bulk of the work is done by `strace`. Re-purposing a shell script as a "library" that another Python script will call is just too spaghetti, and maintenance nightmare.

**Comment by sean-smith** [comment]
Closing as it's stale.

**Pull request closed without merging**